File: go/indicators/johnehlers/fractaladaptivemovingaverage/fractaladaptivemovingaverage.go
package fractaladaptivemovingaverage

//nolint: gofumpt
import (
	"fmt"
	"math"
	"sync"
	"time"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
)

// FractalAdaptiveMovingAverage (Ehler's fractal adaptive moving average, FRAMA)
// is an EMA with the smoothing factor, α, being changed with each new sample:
//
// FRAMAᵢ = αᵢPᵢ + (1 - αᵢ)*FRAMAᵢ₋₁,  αs ≤ αᵢ ≤ 1
//
// Here the αs is the slowest α (default suggested value is 0.01 or equivalent
// length of 199 samples).
//
// The concept of FRAMA is to relate the fractal dimension FDᵢ, calculated on a window
// samples, to the EMA smoothing factor αᵢ, thus making the EMA adaptive.
//
// This dependency is defined as follows:
//
//	αᵢ = exp(-w(FDᵢ - 1)),  1 ≤ FDᵢ ≤ 2,
//
//	w = ln(αs)
//
//	or, given the length ℓs = 2/αs - 1,
//
//	w = ln(2/(ℓs + 1))
//
// The fractal dimension varies over the range from 1 to 2.
//
// When FDᵢ = 1 (series forms a straight line), the exponent is zero – which means that
// αᵢ = 1, and the output of the exponential moving average is equal to the input.
//
// When FDᵢ = 2 (series fills all plane, excibiting extreme volatility), the exponent
// is -w, which means that αᵢ = αs, and the output of the exponential moving average
// is equal to the output of the slowest moving average with αs.
//
// The fractal dimension is estimated by using a "box count" method.
// Since price samples are typically uniformly spaced, the box count is approximated
// by the average slope of the price curve. This is calculated as the highest price
// minus the lowest price within an interval, divided by the length of that interval.
//
// FDᵢ = (ln(N1+N2) − ln(N3)) / ln(2)
//
// N1 is calculated over the first half of the total lookback period ℓ as the
// (highest price - lowest price) during the first ℓ/2 bars, divided by ℓ/2.
//
// N2 is calculated over the second half of the total lookback period ℓ as the
// (lighest price - lowest price) during the second ℓ/2 bars (from ℓ/2 to ℓ-1 bars ago),
// divided by ℓ/2.
//
// N3 is calculated over the entire lookback period ℓ as the
// (highest price - lowest price) during the full ℓ bars, divided by ℓ.
//
// Reference:
//
//	Falconer, K. (2014). Fractal geometry: Mathematical foundations and applications (3rd ed) Wiley.
//	Ehlers, John F. (2005). Fractal Adaptive Moving Average. Technical Analysis of Stocks & Commodities, 23(10), 81–82.
//	Ehlers, John F. (2006). FRAMA – Fractal Adaptive Moving Average, https://www.mesasoftware.com/papers/FRAMA.pdf.
type FractalAdaptiveMovingAverage struct {
	mu               sync.RWMutex
	mnemonic         string
	description      string
	mnemonicFdim     string
	descriptionFdim  string
	alphaSlowest     float64
	scalingFactor    float64
	fractalDimension float64
	value            float64
	length           int
	lengthMinOne     int
	halfLength       int
	windowCount      int
	windowHigh       []float64
	windowLow        []float64
	primed           bool
	barFunc          entities.BarFunc
	quoteFunc        entities.QuoteFunc
	tradeFunc        entities.TradeFunc
}

// NewFractalAdaptiveMovingAverage returns an instance of the indicator
// created using supplied parameters.
func NewFractalAdaptiveMovingAverage( //nolint:funlen
	params *Params,
) (*FractalAdaptiveMovingAverage, error) {
	const (
		invalid = "invalid fractal adaptive moving average parameters"
		fmtl    = "%s: length should be an even integer larger than 1"
		fmta    = "%s: slowest smoothing factor should be in range [0, 1]"
		fmtw    = "%s: %w"
		fmtn    = "frama(%d, %.3f%s)"
		fmtnd   = "framaDim(%d, %.3f%s)"
		two     = 2
		descr   = "Fractal adaptive moving average "
	)

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if params.Length < two {
		return nil, fmt.Errorf(fmtl, invalid)
	}

	if params.SlowestSmoothingFactor < 0. || params.SlowestSmoothingFactor > 1. {
		return nil, fmt.Errorf(fmta, invalid)
	}

	length := params.Length
	if length%2 != 0 {
		length++
	}

	// Resolve defaults for component functions.
	// A zero value means "use default, don't show in mnemonic".
	bc := params.BarComponent
	qc := params.QuoteComponent
	tc := params.TradeComponent

	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	componentMnemonic := core.ComponentTripleMnemonic(bc, qc, tc)
	mnemonic := fmt.Sprintf(fmtn, length, params.SlowestSmoothingFactor, componentMnemonic)
	mnemonicFdim := fmt.Sprintf(fmtnd, length, params.SlowestSmoothingFactor, componentMnemonic)

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	return &FractalAdaptiveMovingAverage{
		mnemonic:         mnemonic,
		description:      descr + mnemonic,
		mnemonicFdim:     mnemonicFdim,
		descriptionFdim:  descr + mnemonicFdim,
		length:           length,
		lengthMinOne:     length - 1,
		halfLength:       length / two,
		windowHigh:       make([]float64, length),
		windowLow:        make([]float64, length),
		windowCount:      0,
		alphaSlowest:     params.SlowestSmoothingFactor,
		scalingFactor:    math.Log(params.SlowestSmoothingFactor),
		fractalDimension: math.NaN(),
		value:            math.NaN(),
		primed:           false,
		barFunc:          barFunc,
		quoteFunc:        quoteFunc,
		tradeFunc:        tradeFunc,
	}, nil
}

// IsPrimed indicates whether an indicator is primed.
func (s *FractalAdaptiveMovingAverage) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes an output data of the indicator.
func (s *FractalAdaptiveMovingAverage) Metadata() core.Metadata {
	return core.Metadata{
		Type:        core.FractalAdaptiveMovingAverage,
		Mnemonic:    s.mnemonic,
		Description: s.description,
		Outputs: []outputs.Metadata{
			{
				Kind:        int(Value),
				Type:        outputs.ScalarType,
				Mnemonic:    s.mnemonic,
				Description: s.description,
			},
			{
				Kind:        int(Fdim),
				Type:        outputs.ScalarType,
				Mnemonic:    s.mnemonicFdim,
				Description: s.descriptionFdim,
			},
		},
	}
}

// Update updates the value of the moving average given the next sample.
func (s *FractalAdaptiveMovingAverage) Update(sample, sampleHigh, sampleLow float64) float64 {
	if math.IsNaN(sampleHigh) || math.IsNaN(sampleLow) || math.IsNaN(sample) {
		return math.NaN()
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	if s.primed {
		for i := range s.lengthMinOne {
			j := i + 1
			s.windowHigh[i] = s.windowHigh[j]
			s.windowLow[i] = s.windowLow[j]
		}

		s.windowHigh[s.lengthMinOne] = sampleHigh
		s.windowLow[s.lengthMinOne] = sampleLow

		s.fractalDimension = s.estimateFractalDimension()
		alpha := s.estimateAlpha()
		s.value += (sample - s.value) * alpha

		return s.value
	} else {
		s.windowHigh[s.windowCount] = sampleHigh
		s.windowLow[s.windowCount] = sampleLow

		s.windowCount++
		if s.windowCount == s.lengthMinOne {
			s.value = sample
		} else if s.windowCount == s.length {
			s.fractalDimension = s.estimateFractalDimension()
			alpha := s.estimateAlpha()
			s.value += (sample - s.value) * alpha
			s.primed = true

			return s.value
		}
	}

	return math.NaN()
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *FractalAdaptiveMovingAverage) UpdateScalar(sample *entities.Scalar) core.Output {
	v := sample.Value

	return s.updateEntity(sample.Time, v, v, v)
}

// UpdateBar updates the indicator given the next bar sample.
func (s *FractalAdaptiveMovingAverage) UpdateBar(sample *entities.Bar) core.Output {
	v := s.barFunc(sample)

	return s.updateEntity(sample.Time, v, sample.High, sample.Low)
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *FractalAdaptiveMovingAverage) UpdateQuote(sample *entities.Quote) core.Output {
	v := s.quoteFunc(sample)

	return s.updateEntity(sample.Time, v, sample.Ask, sample.Bid)
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *FractalAdaptiveMovingAverage) UpdateTrade(sample *entities.Trade) core.Output {
	v := s.tradeFunc(sample)

	return s.updateEntity(sample.Time, v, v, v)
}

func (s *FractalAdaptiveMovingAverage) updateEntity(
	time time.Time, sample, sampleHigh, sampleLow float64,
) core.Output {
	const length = 2

	output := make([]any, length)
	frama := s.Update(sample, sampleHigh, sampleLow)

	fdim := s.fractalDimension
	if math.IsNaN(frama) {
		fdim = math.NaN()
	}

	i := 0
	output[i] = entities.Scalar{Time: time, Value: frama}
	i++
	output[i] = entities.Scalar{Time: time, Value: fdim}

	return output
}

func (s *FractalAdaptiveMovingAverage) estimateFractalDimension() float64 {
	minLowHalf := math.MaxFloat64
	maxHighHalf := math.SmallestNonzeroFloat64

	for i := range s.halfLength {
		l := s.windowLow[i]
		if minLowHalf > l {
			minLowHalf = l
		}

		h := s.windowHigh[i]
		if maxHighHalf < h {
			maxHighHalf = h
		}
	}

	rangeN1 := maxHighHalf - minLowHalf
	minLowFull := minLowHalf
	maxHighFull := maxHighHalf
	minLowHalf = math.MaxFloat64
	maxHighHalf = math.SmallestNonzeroFloat64

	for j := range s.halfLength {
		i := j + s.halfLength
		l := s.windowLow[i]

		if minLowFull > l {
			minLowFull = l
		}

		if minLowHalf > l {
			minLowHalf = l
		}

		h := s.windowHigh[i]
		if maxHighFull < h {
			maxHighFull = h
		}

		if maxHighHalf < h {
			maxHighHalf = h
		}
	}

	rangeN2 := maxHighHalf - minLowHalf
	rangeN3 := maxHighFull - minLowFull

	fdim := (math.Log((rangeN1+rangeN2)/float64(s.halfLength)) -
		math.Log(rangeN3/float64(s.length))) * math.Log2E

	const two = 2

	return math.Min(math.Max(fdim, 1), two)
}

func (s *FractalAdaptiveMovingAverage) estimateAlpha() float64 {
	factor := s.scalingFactor

	// We use the fractal dimension to dynamically change the alpha of an exponential moving average.
	// The fractal dimension varies over the range from 1 to 2.
	// Since the prices are log-normal, it seems reasonable to use an exponential function to relate
	// the fractal dimension to alpha.

	// An empirically chosen scaling in Ehlers's method to map fractal dimension (1–2)
	// to the exponential α.
	alpha := math.Exp(factor * (s.fractalDimension - 1))

	// When the fractal dimension is 1, the exponent is zero – which means that alpha is 1, and
	// the output of the exponential moving average is equal to the input.

	// Limit alpha to vary only from αs to 1.
	return math.Min(math.Max(alpha, s.alphaSlowest), 1)
}

File: ts/indicators/john-ehlers/fractal-adaptive-moving-average/fractal-adaptive-moving-average.ts
import { buildMetadata } from '../../core/build-metadata';
import { Bar } from '../../../entities/bar';
import { BarComponent, DefaultBarComponent, barComponentValue } from '../../../entities/bar-component';
import { Quote } from '../../../entities/quote';
import { QuoteComponent, DefaultQuoteComponent, quoteComponentValue } from '../../../entities/quote-component';
import { Scalar } from '../../../entities/scalar';
import { Trade } from '../../../entities/trade';
import { TradeComponent, DefaultTradeComponent, tradeComponentValue } from '../../../entities/trade-component';
import { Indicator } from '../../core/indicator';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { FractalAdaptiveMovingAverageParams } from './params';

/** __Fractal Adaptive Moving Average__ (Ehler's fractal adaptive moving average, _FRAMA_)
 * is an EMA with the smoothing factor, a, being changed with each new sample:
 *
 *	FRAMAi = aiPi + (1 - ai)*FRAMAi-1,  as <= ai <= 1
 *
 * Here the as is the slowest a (default suggested value is 0.01 or equivalent length of 199 samples).
 *
 * The concept of _FRAMA_ is to relate the fractal dimension FDi, calculated on window
 * samples, to the EMA smoothing factor ai, thus making the EMA adaptive.
 *
 * This dependency is defined as follows:
 *
 *	ai = exp(w(FDi - 1)),  1 <= FDi <= 2,
 *
 *	w = ln(as)
 *
 *	or, given the length ls = 2/as - 1,
 *
 *	w = ln(2/(ls + 1))
 *
 * The fractal dimension varies over the range from 1 to 2.
 *
 * When FDi = 1 (series forms a straight line), the exponent is zero - which means that
 * ai = 1, and the output of the exponential moving average is equal to the input.
 *
 * When FDi = 2 (series fills all plane, excibiting extreme volatility), the exponent
 * is w, which means that ai = as, and the output of the exponential moving average
 * is equal to the output of the slowest moving average with as.
 *
 * The fractal dimension is estimated by using a "box count" method (Falconer 2014 chapter 2).
 *
 *	FDi = (ln(N1+N2) - ln(N3)) / ln(2)
 *
 * Reference:
 *
 *	Falconer, K. (2014). Fractal geometry: Mathematical foundations and applications (3rd ed) Wiley.
 *	Ehlers, John F. (2005). Fractal Adaptive Moving Average. Technical Analysis of Stocks & Commodities, 23(10), 81-82.
 *	Ehlers, John F. (2006). FRAMA - Fractal Adaptive Moving Average, https://www.mesasoftware.com/papers/FRAMA.pdf.
 */
export class FractalAdaptiveMovingAverage implements Indicator {
  private readonly length: number;
  private readonly lengthMinOne: number;
  private readonly halfLength: number;
  private readonly alphaSlowest: number;
  private readonly scalingFactor: number;
  private windowCount: number = 0;
  private value: number = Number.NaN;
  private fractalDimension: number = Number.NaN;
  private readonly windowHigh: number[];
  private readonly windowLow: number[];
  private mnemonic: string;
  private description: string;
  private mnemonicFdim: string;
  private descriptionFdim: string;
  private primed: boolean = false;

  private readonly barComponentFunc: (bar: Bar) => number;
  private readonly quoteComponentFunc: (quote: Quote) => number;
  private readonly tradeComponentFunc: (trade: Trade) => number;

  /**
   * Constructs an instance given input parameters.
   */
  public constructor(params: FractalAdaptiveMovingAverageParams) {
    let len = Math.floor(params.length);
    if (len < 2) {
      throw new Error('length should be an even integer larger than 1');
    }

    if (len % 2 !== 0) {
      len++;
    }

    const alpha = params.slowestSmoothingFactor;
    if (alpha < 0 || alpha > 1) {
      throw new Error('slowest smoothing factor should be in range [0, 1]');
    }

    this.length = len;
    this.lengthMinOne = len - 1;
    this.halfLength = Math.floor(len / 2);
    this.alphaSlowest = alpha;
    this.scalingFactor = Math.log(alpha);
    this.windowHigh = new Array(len).fill(0);
    this.windowLow = new Array(len).fill(0);

    // Resolve component defaults and create component functions.
    const bc = params.barComponent ?? DefaultBarComponent;
    const qc = params.quoteComponent ?? DefaultQuoteComponent;
    const tc = params.tradeComponent ?? DefaultTradeComponent;

    this.barComponentFunc = barComponentValue(bc);
    this.quoteComponentFunc = quoteComponentValue(qc);
    this.tradeComponentFunc = tradeComponentValue(tc);

    // Build mnemonics matching Go format: frama(%d, %.3f%s)
    const cm = componentTripleMnemonic(params.barComponent, params.quoteComponent, params.tradeComponent);
    this.mnemonic = `frama(${len}, ${alpha.toFixed(3)}${cm})`;
    this.mnemonicFdim = `framaDim(${len}, ${alpha.toFixed(3)}${cm})`;

    const descr = 'Fractal adaptive moving average ';
    this.description = descr + this.mnemonic;
    this.descriptionFdim = descr + this.mnemonicFdim;
  }

  /** Indicates whether an indicator is primed. */
  public isPrimed(): boolean { return this.primed; }

  /** Describes a requested output data of an indicator. */
  public metadata(): IndicatorMetadata {
    return buildMetadata(
      IndicatorIdentifier.FractalAdaptiveMovingAverage,
      this.mnemonic,
      this.description,
      [
        { mnemonic: this.mnemonic, description: this.description },
        { mnemonic: this.mnemonicFdim, description: this.descriptionFdim },
      ],
    );
  }

  /** Updates an indicator given the next scalar sample. */
  public updateScalar(sample: Scalar): IndicatorOutput {
    const v = sample.value;

    return this.updateEntity(sample.time, v, v, v);
  }

  /** Updates an indicator given the next bar sample. */
  public updateBar(sample: Bar): IndicatorOutput {
    const v = this.barComponentFunc(sample);

    return this.updateEntity(sample.time, v, sample.high, sample.low);
  }

  /** Updates an indicator given the next quote sample. */
  public updateQuote(sample: Quote): IndicatorOutput {
    const v = this.quoteComponentFunc(sample);

    return this.updateEntity(sample.time, v, sample.askPrice, sample.bidPrice);
  }

  /** Updates an indicator given the next trade sample. */
  public updateTrade(sample: Trade): IndicatorOutput {
    const v = this.tradeComponentFunc(sample);

    return this.updateEntity(sample.time, v, v, v);
  }

  /** Updates the value of the indicator given the next sample with high and low values. */
  public update(sample: number, sampleHigh: number, sampleLow: number): number {
    if (Number.isNaN(sample) || Number.isNaN(sampleHigh) || Number.isNaN(sampleLow)) {
      return Number.NaN;
    }

    if (this.primed) {
      for (let i = 0; i < this.lengthMinOne; i++) {
        const j = i + 1;
        this.windowHigh[i] = this.windowHigh[j];
        this.windowLow[i] = this.windowLow[j];
      }

      this.windowHigh[this.lengthMinOne] = sampleHigh;
      this.windowLow[this.lengthMinOne] = sampleLow;

      this.fractalDimension = this.estimateFractalDimension();
      const a = this.estimateAlpha();
      this.value += (sample - this.value) * a;

      return this.value;
    }

    this.windowHigh[this.windowCount] = sampleHigh;
    this.windowLow[this.windowCount] = sampleLow;

    this.windowCount++;
    if (this.windowCount === this.lengthMinOne) {
      this.value = sample;
    } else if (this.windowCount === this.length) {
      this.fractalDimension = this.estimateFractalDimension();
      const a = this.estimateAlpha();
      this.value += (sample - this.value) * a;
      this.primed = true;

      return this.value;
    }

    return Number.NaN;
  }

  private updateEntity(time: Date, sample: number, sampleHigh: number, sampleLow: number): IndicatorOutput {
    const frama = this.update(sample, sampleHigh, sampleLow);

    let fdim = this.fractalDimension;
    if (Number.isNaN(frama)) {
      fdim = Number.NaN;
    }

    const scalarFrama = new Scalar();
    scalarFrama.time = time;
    scalarFrama.value = frama;

    const scalarFdim = new Scalar();
    scalarFdim.time = time;
    scalarFdim.value = fdim;

    return [scalarFrama, scalarFdim];
  }

  private estimateFractalDimension(): number {
    let minLowHalf = Number.MAX_VALUE;
    let maxHighHalf = -Number.MAX_VALUE;

    for (let i = 0; i < this.halfLength; i++) {
      const l = this.windowLow[i];
      if (minLowHalf > l) {
        minLowHalf = l;
      }

      const h = this.windowHigh[i];
      if (maxHighHalf < h) {
        maxHighHalf = h;
      }
    }

    const rangeN1 = maxHighHalf - minLowHalf;
    let minLowFull = minLowHalf;
    let maxHighFull = maxHighHalf;
    minLowHalf = Number.MAX_VALUE;
    maxHighHalf = -Number.MAX_VALUE;

    for (let i = this.halfLength; i < this.length; i++) {
      const l = this.windowLow[i];
      if (minLowHalf > l) {
        minLowHalf = l;
      }
      if (minLowFull > l) {
        minLowFull = l;
      }

      const h = this.windowHigh[i];
      if (maxHighHalf < h) {
        maxHighHalf = h;
      }
      if (maxHighFull < h) {
        maxHighFull = h;
      }
    }

    const rangeN2 = maxHighHalf - minLowHalf;
    const rangeN3 = maxHighFull - minLowFull;

    const fdim = (Math.log((rangeN1 + rangeN2) / this.halfLength) -
      Math.log(rangeN3 / this.length)) * Math.LOG2E;

    return Math.min(Math.max(fdim, 1), 2);
  }

  private estimateAlpha(): number {
    const factor = this.scalingFactor;

    // We use the fractal dimension to dynamically change the alpha of an exponential moving average.
    // The fractal dimension varies over the range from 1 to 2.
    // Since the prices are log-normal, it seems reasonable to use an exponential function to relate
    // the fractal dimension to alpha.

    // An empirically chosen scaling in Ehlers's method to map fractal dimension (1-2)
    // to the exponential a.
    const alpha = Math.exp(factor * (this.fractalDimension - 1));

    // When the fractal dimension is 1, the exponent is zero - which means that alpha is 1, and
    // the output of the exponential moving average is equal to the input.

    // Limit alpha to vary only from as to 1.
    return Math.min(Math.max(alpha, this.alphaSlowest), 1);
  }
}
